src/mcts/search.cc

```diff
@@ -556,18 +567,38 @@ void SearchWorker::GatherMinibatch() {
       if (++collisions_found > search_->kAllowedNodeCollisions) return;
       continue;
     }
-    ++nodes_found;
+    ++minibatch_size;
+
     // If node is already known as terminal (win/loss/draw according to rules
     // of the game), it means that we already visited this node before.
-    if (node->IsTerminal()) continue;
+    if (!node->IsTerminal()) {
+      // Node was never visited, extend it.
+      ExtendNode(node);
 
-    // Node was never visited, extend it.
-    ExtendNode(node);
+      // Only send non-terminal nodes to neural network
+      if (!node->IsTerminal()) {
+        minibatch_.back().nn_queried = true;
+        minibatch_.back().is_cache_hit = AddNodeToComputation(node, true);
+      }
+    }
 
-    // Only send non-terminal nodes to neural network
-    if (!node->IsTerminal()) {
-      nodes_to_process_.back().nn_queried = true;
-      AddNodeToComputation(node);
+    // If out of order eval is enabled and the node to compute we added last
+    // doesn't require NN eval (i.e. it's a cache hit or terminal node), do
+    // out of order eval for it.
+    if (search_->kOutOfOrderEval) {
+      if (node->IsTerminal() || minibatch_.back().is_cache_hit) {
+        // Perform out of order eval for the last entry in minibatch_.
+        FetchSingleNodeResult(&minibatch_.back(),
+                              computation_->GetBatchSize() - 1);
+        DoBackupUpdateSingleNode(minibatch_.back());
+
+        // Remove last entry in minibatch_, as it has just been
+        // processed.
+        // If NN eval was already processed out of order, remove it.
+        if (minibatch_.back().nn_queried) computation_->PopCacheHit();
+        minibatch_.pop_back();
+        --minibatch_size;
```

Comment:
Ditto this block as well about the local variable!

Reply:
Done.